[serve] add script for comparing perf metrics

zcin opened the pull request:
[serve] add script for comparing perf metrics

Add script that can:
* output the comparison between the performance metrics from two release test output files.
* fetch release test results from latest nightly run on master

Usage:
```
./release_utils.py compare-perf RESULTS_FILE1 RESULTS_FILE2
```

Example output:
```
|metric                        |results 1      |results 2      |%change   |
|------------------------------|---------------|---------------|----------|
|http_p50_latency              |11.08          |11.63          |4.9       |
|http_p90_latency              |17.24          |17.71          |2.75      |
|http_p95_latency              |19.19          |20.12          |4.87      |
|http_p99_latency              |22.7           |25.29          |11.44     |
|http_1mb_p50_latency          |11.82          |12.78          |8.12      |
|http_1mb_p90_latency          |18.78          |19.59          |4.32      |
|http_1mb_p95_latency          |21.55          |22.79          |5.75      |
|http_1mb_p99_latency          |32.36          |26.53          |-18.0     |
|http_10mb_p50_latency         |17.57          |18.04          |2.65      |
|http_10mb_p90_latency         |25.32          |26.31          |3.88      |
|http_10mb_p95_latency         |28.41          |30.15          |6.14      |
|http_10mb_p99_latency         |35.19          |37.68          |7.08      |
|http_avg_rps                  |204.2          |195.04         |-4.49     |
|http_throughput_std           |10.59          |10.05          |-5.1      |
|grpc_p50_latency              |7.97           |8.84           |11.03     |
|grpc_p90_latency              |12.89          |16.0           |24.14     |
|grpc_p95_latency              |15.1           |19.05          |26.21     |
|grpc_p99_latency              |21.11          |25.19          |19.3      |
|grpc_1mb_p50_latency          |17.65          |19.92          |12.85     |
|grpc_1mb_p90_latency          |24.59          |28.3           |15.1      |
|grpc_1mb_p95_latency          |27.62          |32.11          |16.27     |
|grpc_1mb_p99_latency          |34.33          |40.01          |16.57     |
|grpc_10mb_p50_latency         |142.4          |153.89         |8.07      |
|grpc_10mb_p90_latency         |171.29         |196.54         |14.74     |
|grpc_10mb_p95_latency         |180.63         |208.3          |15.32     |
|grpc_10mb_p99_latency         |203.08         |228.68         |12.61     |
|grpc_avg_rps                  |203.35         |211.01         |3.77      |
|grpc_throughput_std           |11.9           |18.37          |54.37     |
|handle_p50_latency            |4.89           |4.08           |-16.52    |
|handle_p90_latency            |8.33           |7.07           |-15.17    |
|handle_p95_latency            |10.29          |8.81           |-14.38    |
|handle_p99_latency            |14.3           |11.08          |-22.5     |
|handle_1mb_p50_latency        |11.58          |10.91          |-5.85     |
|handle_1mb_p90_latency        |17.04          |16.73          |-1.86     |
|handle_1mb_p95_latency        |19.41          |18.97          |-2.29     |
|handle_1mb_p99_latency        |22.92          |26.54          |15.78     |
|handle_10mb_p50_latency       |65.54          |67.52          |3.02      |
|handle_10mb_p90_latency       |83.53          |84.07          |0.65      |
|handle_10mb_p95_latency       |90.7           |93.42          |3.0       |
|handle_10mb_p99_latency       |106.93         |104.65         |-2.13     |
|handle_avg_rps                |394.57         |404.85         |2.61      |
|handle_throughput_std         |22.09          |23.86          |8.01      |
```

The output can also be directly copy-pasted into a github PR and it will be formatted as a human-readable table. e.g. the above becomes:

|metric                        |results 1      |results 2      |%change   |
|------------------------------|---------------|---------------|----------|
|http_p50_latency              |11.08          |11.63          |4.9       |
|http_p90_latency              |17.24          |17.71          |2.75      |
|http_p95_latency              |19.19          |20.12          |4.87      |
|http_p99_latency              |22.7           |25.29          |11.44     |
|http_1mb_p50_latency          |11.82          |12.78          |8.12      |
|http_1mb_p90_latency          |18.78          |19.59          |4.32      |
|http_1mb_p95_latency          |21.55          |22.79          |5.75      |
|http_1mb_p99_latency          |32.36          |26.53          |-18.0     |
|http_10mb_p50_latency         |17.57          |18.04          |2.65      |
|http_10mb_p90_latency         |25.32          |26.31          |3.88      |
|http_10mb_p95_latency         |28.41          |30.15          |6.14      |
|http_10mb_p99_latency         |35.19          |37.68          |7.08      |
|http_avg_rps                  |204.2          |195.04         |-4.49     |
|http_throughput_std           |10.59          |10.05          |-5.1      |
|grpc_p50_latency              |7.97           |8.84           |11.03     |
|grpc_p90_latency              |12.89          |16.0           |24.14     |
|grpc_p95_latency              |15.1           |19.05          |26.21     |
|grpc_p99_latency              |21.11          |25.19          |19.3      |
|grpc_1mb_p50_latency          |17.65          |19.92          |12.85     |
|grpc_1mb_p90_latency          |24.59          |28.3           |15.1      |
|grpc_1mb_p95_latency          |27.62          |32.11          |16.27     |
|grpc_1mb_p99_latency          |34.33          |40.01          |16.57     |
|grpc_10mb_p50_latency         |142.4          |153.89         |8.07      |
|grpc_10mb_p90_latency         |171.29         |196.54         |14.74     |
|grpc_10mb_p95_latency         |180.63         |208.3          |15.32     |
|grpc_10mb_p99_latency         |203.08         |228.68         |12.61     |
|grpc_avg_rps                  |203.35         |211.01         |3.77      |
|grpc_throughput_std           |11.9           |18.37          |54.37     |
|handle_p50_latency            |4.89           |4.08           |-16.52    |
|handle_p90_latency            |8.33           |7.07           |-15.17    |
|handle_p95_latency            |10.29          |8.81           |-14.38    |
|handle_p99_latency            |14.3           |11.08          |-22.5     |
|handle_1mb_p50_latency        |11.58          |10.91          |-5.85     |
|handle_1mb_p90_latency        |17.04          |16.73          |-1.86     |
|handle_1mb_p95_latency        |19.41          |18.97          |-2.29     |
|handle_1mb_p99_latency        |22.92          |26.54          |15.78     |
|handle_10mb_p50_latency       |65.54          |67.52          |3.02      |
|handle_10mb_p90_latency       |83.53          |84.07          |0.65      |
|handle_10mb_p95_latency       |90.7           |93.42          |3.0       |
|handle_10mb_p99_latency       |106.93         |104.65         |-2.13     |
|handle_avg_rps                |394.57         |404.85         |2.61      |
|handle_throughput_std         |22.09          |23.86          |8.01      |

Example usage:
```
./release_utils.py fetch-nightly TEST_NAME -o OUTPUT_PATH
```

I also added options to the release tests `microbenchmarks.py`, `autoscaling_load_test.py`, and `replica_scalability.py` so that they can be more easily run locally, and the release test output can be saved to a filepath of your choice with `-o output_path`. This can be helpful if we want to quickly run the microbenchmarks locally in the future and compare against some other results.

Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

edoakes commented:
Is it possible to get the results file from the latest nightly run somehow?

Or could we even add logic to the script that fetches the latest persisted output?

zcin commented:
@edoakes That seems useful! I've added to the script and renamed it to `release_utils.py`. Example usage:

```
> ./release_utils.py fetch-nightly serve_autoscaling_load_test.aws -o ~/workspace/comp2
INFO:/Users/cindyz/ray/release/serve_tests/./release_utils.py:Found latest release run on master with build #14173 and job ID 018f115d-836d-4547-8ca6-1ebc936930d6.

> ./release_utils.py compare-perf ~/workspace/comp1 ~/workspace/comp2
|metric                        |results 1      |results 2      |%change   |
|------------------------------|---------------|---------------|----------|
|p50_latency                   |830            |760            |-8.43     |
|p90_latency                   |1500           |1400           |-6.67     |
|p99_latency                   |2000           |1900           |-5.0      |
|avg_rps                       |58.8           |62.94          |7.03      |
|stage_1_p50_latency           |690            |660            |-4.35     |
|stage_1_p90_latency           |1100           |1200           |9.09      |
|stage_1_p99_latency           |1400           |1600           |14.29     |
|stage_1_rps                   |14.2           |14.6           |2.82      |
|stage_2_p50_latency           |840            |790            |-5.95     |
|stage_2_p90_latency           |1500           |1400           |-6.67     |
|stage_2_p99_latency           |1800           |1800           |0.0       |
|stage_2_rps                   |55.7           |56.4           |1.26      |
|stage_3_p50_latency           |870            |820            |-5.75     |
|stage_3_p90_latency           |1500           |1400           |-6.67     |
|stage_3_p99_latency           |1800           |1800           |0.0       |
|stage_3_rps                   |108.7          |116.1          |6.81      |
```
